driver-core/src/main/com/mongodb/connection/ScramShaAuthenticator.java

```diff
@@ -0,0 +1,437 @@
+        private int step = -1;
+
+        ScramShaSaslClient(final MongoCredential credential, final AuthenticationMechanism authenticationMechanism,
+                           final RandomStringGenerator randomStringGenerator,
+                           final AuthenticationHashGenerator authenticationHashGenerator) {
+            this.credential = credential;
+            this.randomStringGenerator = randomStringGenerator;
+            this.authenticationHashGenerator = authenticationHashGenerator;
+            this.authenticationMechanism = authenticationMechanism;
+            if (authenticationMechanism.equals(SCRAM_SHA_1)) {
+                hAlgorithm = "SHA-1";
+                hmacAlgorithm = "HmacSHA1";
+            } else {
+                hAlgorithm = "SHA-256";
+                hmacAlgorithm = "HmacSHA256";
+            }
+        }
+
+        public String getMechanismName() {
+            return authenticationMechanism.getMechanismName();
+        }
+
+        public boolean hasInitialResponse() {
+            return true;
+        }
+
+        public byte[] evaluateChallenge(final byte[] challenge) throws SaslException {
+            step++;
+            if (step == 0) {
+                return computeClientFirstMessage();
+            } else if (step == 1) {
+                return computeClientFinalMessage(challenge);
+            } else if (step == 2) {
+                return validateServerSignature(challenge);
+            } else {
+                throw new SaslException(format("Too many steps involved in the %s negotiation.", getMechanismName()));
+            }
+        }
+
+        private byte[] validateServerSignature(final byte[] challenge) throws SaslException {
+            String serverResponse = encodeUTF8(challenge);
+            HashMap<String, String> map = parseServerResponse(serverResponse);
+            if (!MessageDigest.isEqual(decodeBase64(map.get("v")), serverSignature)) {
+                throw new SaslException("Server signature was invalid.");
+            }
+            return challenge;
+        }
+
+        public boolean isComplete() {
+            return step == 3;
+        }
+
+        public byte[] unwrap(final byte[] incoming, final int offset, final int len) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        public byte[] wrap(final byte[] outgoing, final int offset, final int len) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        public Object getNegotiatedProperty(final String propName) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        public void dispose() {
+            // nothing to do
+        }
+
+        private byte[] computeClientFirstMessage() throws SaslException {
+            clientNonce = randomStringGenerator.generate(RANDOM_LENGTH);
+
+            StringWriter writer = new StringWriter();
```

Comment:
Use StringBuilder not StringWriter.  See https://stackoverflow.com/questions/2980805/string-assembly-by-stringbuilder-vs-stringwriter-and-printwriter

Reply:
Done - intellij tells me += is actually as performant to StringBuilder